### PR TITLE
Add nonce support to the oauth2 example verifications

### PR DESCRIPTION
**Pull request by zhaohuabing**
This PR adds nonce support to the oauth2 example verifications to allow the PR https://github.com/envoyproxy/envoy/pull/35919 to pass.

@phlax

### CODE REVIEW

File: single-page-app/envoy.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ static_resources:
                   uri: http://myhub:7000/authenticate
                   timeout: 3s
                 default_expires_in: 600s
-                authorization_endpoint: http://localhost:7000/authorize
+                authorization_endpoint: http://localhost:11902/authorize
```
Comment (phlax) [inline single-page-app/envoy.yml:35]: this is incorrect - please keep original - same below

**Review by phlax — COMMENTED** [review]
this is more/less disabling the entire test - not sure im comfortable with that, but if we are going to do it use `return` rather than commenting everything out

**Comment by phlax** [comment]
@zhaohuabing i more meant fix the test so that it will work when the PR lands rather than disabling everything so it can pass now

**Comment by zhaohuabing** [comment]
> @zhaohuabing i more meant fix the test so that it will work when the PR lands rather than disabling everything so it can pass now

I fixed the tests "Inititiate login" and "Fetch the myhub authorization page", but I can't make "Return to the app and receive creds" and "Log out of Myhub" work for both versions of OAuth2 implementations. The verification of state in the OAuth2 filter change:
* The current version: It verify the state is a valid URL representing the original request
* The new version: in addition to the URL validation, the OAuth2 filter also validate the nonce

**Comment by phlax** [comment]
> ... the OAuth2 filter also validates the nonce ...

yeah im not sure that `responds_with_headers` it going to be able to handle that out of the box

i can think of some slightly more complex patterns that would - eg fetch the response header, parse the nonce, interpolate that in the `responds_with_header` test

**Comment by zhaohuabing** [comment]
Ha, I can probably add an `if else` logic in the test to accommodate both the old and new implementations.

**Comment by phlax** [comment]
checking `verify-common.sh` - i think `responds_with_header` should work iff the nonce is added at the end - it does a simple grep

https://github.com/envoyproxy/examples/blob/86c99243f0c4403bc234f5a6763f4654173d5c44/verify-common.sh#L190-L203

but yeah - repeating what i said offline - im more interested in having a fix ready than disabling stuff

**Comment by phlax** [comment]
if/else probs does make sense because end users may have either container locally so would make it more robust

**Comment by phlax** [comment]
hmm - and now im remembering that the examples are "docs-first" and the verify script is really just for ci - so probs if/else doesnt matter too much

what does matter tho is that we update the docs if anything changed there (the tests are meant to ~mirror the docs so i expect so)

File: single-page-app/verify.sh [hunk 2]
```diff
@@ -118,29 +118,63 @@ test_auth () {
         "${proxy_scheme}://localhost:${proxy_port}" \
         "${curl_args[@]}"
 
+    run_log "Check whether the nonce is used in the OAuth2 filter"
+    SUPPORT_NONCE="false"
+    LOCATION=$(_curl "${curl_args[@]}" --head "${proxy_scheme}://localhost:${proxy_port}/login" | grep location)
+    if [[ "$LOCATION" == *"nonce%3D"* ]]; then
+        SUPPORT_NONCE="true"
```
Comment (phlax) [inline single-page-app/verify.sh:125]: lets add a comment here about removal - not sure when - maybe a year?

**Review by phlax — COMMENTED** [review]
looks good @zhaohuabing thanks!

if we can test it in envoy now - then im happy to land it - thanks again

**Comment by zhaohuabing** [comment]
> looks good @zhaohuabing thanks!
> 
> if we can test it in envoy now - then im happy to land it - thanks again

I already tested it locally with https://github.com/envoyproxy/envoy/pull/35919 and the verification passed.

Can we link this PR with the verification CI task of https://github.com/envoyproxy/envoy/pull/35919?

**Comment by phlax** [comment]
yeah you should be able to use the commit hash in `bazel/repository_locations.bzl` 

alternately - and perhaps the better approach - is to add a patch for this in envoy

**Review by phlax — APPROVED** [review]
lgtm, thanks a lot for addressing this properly @zhaohuabing - really appreciated

ill try and follow up to add some docs somewhere on the process for adding examples-breaking changes - this was new ground

**Comment by phlax** [comment]
given this should work with old/new code - im thinking we can probably just cut a release and update envoy immediately

**Pull request merged**